src/proc/mod.rs

```diff
@@ -108,6 +118,37 @@ impl super::TypeInner {
             Self::Image { .. } | Self::Sampler { .. } => 0,
         }
     }
+
+    /// Return the length of a subscriptable type, if known at compile time.
+    ///
+    /// The value returned is appropriate for bounds checks on subscripting. If
+    /// the length is dynamic (say, a WGSL runtime-sized array), then return
+    /// `None`.
+    ///
+    /// `self` must be a vector, matrix, or array.
+    pub fn to_known_length(&self, module: &super::Module) -> Option<u32> {
+        match *self {
+            Self::Vector { size, .. } => Some(size as _),
+            Self::Matrix { columns, .. } => Some(columns as _),
+            Self::Array { size, .. } => size.to_known_length(module),
+            Self::ValuePointer {
+                size: Some(size), ..
+            } => Some(size as _),
+            Self::Pointer { base, .. } => {
+                // When assigning types to expressions, ResolveContext::Resolve
+                // does a separate sub-match here instead of a full recursion,
+                // so we'll do the same.
+                let base_inner = &module.types[base].inner;
+                match *base_inner {
+                    Self::Vector { size, .. } => Some(size as _),
+                    Self::Matrix { columns, .. } => Some(columns as _),
+                    Self::Array { size, .. } => size.to_known_length(module),
+                    _ => unreachable!("pointed-to type does not have a length: {:?}", base_inner),
```

Comment:
How about return `None` here and add warn log with this message?

Reply:
Hmm, `None` means a dynamically sized array; we take advantage of that in `back/spv/writer.rs`. Maybe this could return `Result<Option<u32>>`?

Reply:
@kvark What approach would you recommend here?

Reply:
Is this supposed to be called by the both the backends and the frontends? If it's just the backends, moving it to `back/mod.rs` and leaving the unreachable is fine.
If it's both, let's return a result

Reply:
It's used in expression validation to do some checks at shader translation time. I'll make it return a `Result`.